src/transform/src/normalize_lets.rs

```diff
@@ -490,9 +498,30 @@ mod inlining {
             for (id, mut expr) in ids.drain(..).zip(values.drain(..)) {
                 // Substitute any appropriate prior let bindings.
                 inline_lets_helper(&mut expr, &mut inline_offer)?;
+
+                // Determine the first `id'` at which any inlining offer must expire.
+                // An inlining offer expires because it references an `id'` that is not yet bound,
+                // indicating a reference to the *next* iterate of that identifier. Inlining the
+                // expression once `id'` becomes bound would advance the reference to be to the
+                // *current* iterate of the identifier.
+                expr.visit_pre(|e| {
+                    if let MirRelationExpr::Get {
+                        id: Id::Local(referenced_id),
+                        ..
+                    } = e
+                    {
+                        if referenced_id >= &id {
+                            expire_offers
+                                .entry(*referenced_id)
+                                .or_insert_with(Vec::new)
+                                .push(id);
+                        }
+                    }
+                });
+
                 // Gets for `id` only occur in later expressions, so this should still be correct.
```

Comment:
> If there is more than one Get for id, then we shouldn't inline, regardless of how many of these are occurring before or after us.
.. unless it's a sufficiently simple expression I think?

Reply:
This comment is inherited from earlier versions of the code, but I think it refers to this concern from further up in the file:
```rust
            //   1. by going in order no substitution will increase the `Get`-count of an
            //      identifier beyond one, as all in values with strictly greater identifiers.
```
Those words don't make total sense to me, but you can end up failing to perform some fairly simple substitutions because the apparent multiplicity of reference increases. I don't have the examples on the tip of my tongue unfortunately. Perhaps some future time we can see what happens if the code is shuffled around, and if we find we understand it better we can update the comment.